### PR TITLE
#918 Fix for forcing LSI when query possible with table's primary index

### LINKED ISSUE #918: [BUG] Wrong query with LocalSecondaryIndex
### Summary:
I'm trying to define schema with rangeKey and LocalSecondaryIndex (see schema). But whenever I do this, query method will generate bad request.

#### Schema
```
new dynamoose.Schema({
        userId: {
            type: String,
            hashKey: true,
            index: {
                name: 'dateTimeIndex',
                global: false,
                rangeKey: 'dateTime'
            }
        },
        createdAt: {
            type: Date,
            rangeKey: true
        },
        dateTime: {
            type: Date
        }
})
```

#### Code
```
this.Meeting.query( 'userId').eq(userId)
            .where('createdAt').ge(0)
            .exec()
```

### Current output and behavior (including stack trace):

```
{
    "FilterExpression": "#a1 >= :v1",
    "ExpressionAttributeNames": {
        "#a1": "createdAt",
        "#qha": "userId"
    },
    "ExpressionAttributeValues": {
        ":v1": {
            "N": "0"
        },
        ":qhv": {
            "S": "user"
        }
    },
    "TableName": "meetings",
    "IndexName": "dateTimeIndex",
    "KeyConditionExpression": "#qha = :qhv"
}
```
There is extra "IndexName" and "FilterExpression".

### Expected output and behavior:
I would like to see something like this instead:
```
{
  "TableName": "meetings",
  "ScanIndexForward": false,
  "ConsistentRead": false,
  "KeyConditionExpression": "#fb330 = :fb330 And #fb331 >= :fb331",
  "ExpressionAttributeValues": {
    ":fb330": {
      "S": "user"
    },
    ":fb331": {
      "N": "0"
    }
  },
  "ExpressionAttributeNames": {
    "#fb330": "userId",
    "#fb331": "createdAt"
  }
}
```
(generated by NoSQL Workbench)

When I change schema after Table creation and remove "index" from hashKey problem disappears.

### Environment:
Operating System: MacOS
Operating System Version: 10.15.4
Node.js version (`node -v`): v10.13.0
NPM version: (`npm -v`): 6.14.4
Dynamoose version: 2.2.0

### Other:
- [x] I have read through the Dynamoose documentation before posting this issue
- [x] I have searched through the GitHub issues (including closed issues) and pull requests to ensure this issue has not already been raised before
- [x] I have searched the internet and Stack Overflow to ensure this issue hasn't been raised or answered before
- [x] I have tested the code provided and am confident it doesn't work as intended
- [x] I have filled out all fields above
- [x] I am running the latest version of Dynamoose

### PR DESCRIPTION
**Pull request by wips**
### Summary:
The issue is described in #918. The only thing that bothers me with this fix is that it'll break those queries which currently may fallback to LSI. Someone may not notice that they forgot to mention the LSI name in query and get error. Though should be fine as soon as the index name is set using `using`.

### GitHub linked issue:
Closes #918 

### Type (select 1):
- [x] Bug fix
- [ ] Feature implementation
- [ ] Documentation improvement
- [ ] Testing improvement

- [ ] Test added to report bug (GitHub issue #--- @---)
- [ ] Something not listed here

### Is this a breaking change? (select 1):
- [x] 🚨 YES 🚨 (if we consider previous behaviour "normal" then yes, otherwise can be considered a bugfix)
- [ ] No
- [ ] I'm not sure

### Is this ready to be merged into Dynamoose? (select 1):
- [x] Yes
- [ ] No

### Are all the tests currently passing on this PR? (select 1):
- [x] Yes
- [ ] No

### Other:
- [x] I have read through and followed the Contributing Guidelines
- [x] I have searched through the GitHub pull requests to ensure this PR has not already been submitted
- [x] I have updated the Dynamoose documentation (if required) given the changes I made
- [x] I have ensured the following commands are successful from the root of the project directory
  - [x] `npm test`
  - [x] `npm run lint`
- [x] I have added/updated the Dynamoose test cases (if required) given the changes I made
- [x] I agree that all changes made in this pull request may be distributed and are made available in accordance with the [Dynamoose license](https://github.com/dynamoose/dynamoose/blob/master/LICENSE)
- [x] All of my commits and commit messages are detailed, explain what changes were made, and are easy to follow and understand
- [x] I have filled out all fields above

### CODE REVIEW

File: lib/DocumentRetriever.ts [hunk 1]
```diff
@@ -157,6 +157,22 @@ DocumentRetriever.prototype.getRequest = async function (this: DocumentRetriever
...
+		const hashKeyInQuery = Object.entries(chart)
+			.find(([fieldName, {type}]) => type === "EQ" && fieldName === hashKeyOfTable);
+		if (!hashKeyInQuery) {
+			return false;
+		}
+		const isOneKeyQuery = Object.keys(chart).length === 1;
+		if (isOneKeyQuery && hashKeyInQuery) {
+			return true;
+		} else {
+			if (rangeKeyOfTable) {
```
Comment (fishcharlie) [inline lib/DocumentRetriever.ts:170]:
```suggestion
		} else if (rangeKeyOfTable) {
```

File: lib/DocumentRetriever.ts [hunk 2]
```diff
@@ -157,6 +157,22 @@ DocumentRetriever.prototype.getRequest = async function (this: DocumentRetriever
 		object.ExclusiveStartKey = Document.isDynamoObject(this.settings.startAt) ? this.settings.startAt : this.internalSettings.model.Document.objectToDynamo(this.settings.startAt);
 	}
 	const indexes = await this.internalSettings.model.getIndexes();
+	function canUseIndexOfTable (hashKeyOfTable, rangeKeyOfTable, chart: ConditionStorageTypeNested): boolean {
+		const hashKeyInQuery = Object.entries(chart)
+			.find(([fieldName, {type}]) => type === "EQ" && fieldName === hashKeyOfTable);
```
Comment (fishcharlie) [inline lib/DocumentRetriever.ts:162]:
```suggestion
		const hashKeyInQuery = Object.entries(chart).find(([fieldName, {type}]) => type === "EQ" && fieldName === hashKeyOfTable);
```

File: test/unit/Query.js [hunk 3]
```diff
@@ -729,6 +729,61 @@ describe("Query", () => {
 				"KeyConditionExpression": "#qha = :qhv"
 			});
 		});
+
+		describe("using index of table", () => {
```
Comment (fishcharlie) [inline test/unit/Query.js:733]:
```suggestion
		describe("Using index of table", () => {
```

**Comment by fishcharlie** [comment]
@sreeparnabanerjee53 @thebrokenbar Could one (or ideally both) of you please review & TEST this PR to make sure it solves your problem?

File: lib/DocumentRetriever.ts [hunk 4]
```diff
@@ -157,6 +157,20 @@ DocumentRetriever.prototype.getRequest = async function (this: DocumentRetriever
...
+		if (!hashKeyInQuery) {
+			return false;
+		}
+		const isOneKeyQuery = Object.keys(chart).length === 1;
+		if (isOneKeyQuery && hashKeyInQuery) {
+			return true;
+		} else if (rangeKeyOfTable) {
+				return Object.entries(chart).some(([fieldName]) => fieldName !== hashKeyInQuery[0]);
+			}
+		}
```
Comment (fishcharlie) [inline lib/DocumentRetriever.ts:171]:
```suggestion
			return Object.entries(chart).some(([fieldName]) => fieldName !== hashKeyInQuery[0]);
		}
```

**Review by fishcharlie — APPROVED** [review]

**Review by sreeparnabanerjee53 — APPROVED** [review]

**Comment by wips** [comment]
> @sreeparnabanerjee53 @thebrokenbar Could one (or ideally both) of you please review & TEST this PR to make sure it solves your problem?

Just for the record - I'm having issue with this too and this PR solves it for me =)

**Comment by fishcharlie** [comment]
@wips lol, because submitting a PR that doesn't fix the issue from your perspective is a great idea 😆

**Pull request merged**